classes/step1_form.php

```diff
@@ -0,0 +1,220 @@
+            $mform->getElement($textname)->setValue($recipe[$textname]);
+        }
+
+        if (!empty($templatevar['required'])) {
+            $mform->addRule($textname, null, 'required', null, 'client', false, true);
+        }
+    }
+
+    /**
+     * Adds a fieldset element to the form.
+     *
+     * @param string[] $templatevar The template variable
+     * @param string[] $recipe The recipe.
+     */
+    protected function add_fieldset($templatevar, $recipe) {
+
+        $mform =& $this->_form;
+        $name = $templatevar['name'];
+        $elements = $templatevar['values'];
+
+        if (empty($this->_customdata[$name.'count'])) {
+            // Create only one entry in the fieldset if more haven't been specified.
+            $count = 1;
+        } else {
+            $count = (int) $this->_customdata[$name.'count'];
+        }
+
+        $recipevalues = array();
+        // Keeping only the recipe values that are part of the template variable.
+        if (!empty($recipe[$name]) && is_array($recipe[$name])) {
+            foreach ($recipe[$name] as $recipevalue) {
+
+                $currentvalue = array();
+                foreach ($elements as $element) {
+                    if (isset($recipevalue[$element])) {
+                        $currentvalue[$element] = $recipevalue[$element];
+                    }
+                }
+
+                if (!empty($currentvalue)) {
+                    $recipevalues[] = $currentvalue;
+                }
+            }
+        }
+
+        $mform->addElement('header', $name, get_string('skel'.$name, 'tool_pluginskel'));
+
+        if (!empty($templatevar['required']) || !empty($recipevalues)) {
+            $mform->setExpanded($name, true);
+        } else {
+            $mform->setExpanded($name, false);
+        }
+
+        if (!empty($recipevalues)) {
+            foreach ($recipevalues as $index => $value) {
+                foreach ($elements as $element) {
+                    $elementname = $name.'['.$index.']['.$element.']';
+                    $mform->addElement('text', $elementname, get_string('skel'.$element, 'tool_pluginskel'));
+                    $mform->setType($elementname, PARAM_TEXT);
+
+                    if (!empty($value[$element])) {
+                        $mform->getElement($elementname)->setValue($value[$element]);
+                    }
+                }
+            }
+        }
+
+        $currentcount = count($recipevalues);
+
+        while ($currentcount < $count) {
+            foreach ($elements as $element) {
+                $elementname = $name.'['.$currentcount.']['.$element.']';
+                $mform->addElement('text', $elementname, get_string('skel'.$element, 'tool_pluginskel'));
+                $mform->setType($elementname, PARAM_TEXT);
+            }
+            $currentcount = $currentcount + 1;
+        }
+
+        $mform->addElement('submit', 'addmore_'.$name, get_string('addmore', 'tool_pluginskel'));
```

Comment:
I haven't implemented this yet. When the user presses the Add more button more text elements will be created on the web page.

The best way to do it would be with Javascript, I'll find a way to do it.